File: apzkr-pzpi-21-11-milennyi-ivan/Task3-WebClient/src/api/services/FeedSupplyService.ts
import axios from "axios";
import { FeedSupply } from "../../models/FeedSupply";
import { API_URL } from "../../utils/config";
import { FeedSupplyVM } from "../../viewModels/FeedSupplyVM";
import { IFeedSupplyService } from "../interfaces/IFeedSupplyService";

export class FeedSupplyService implements IFeedSupplyService {
    // Статичні URL-адреси для запитів до API
    private static FEED_SUPPLY_URLS = {
        GET_ALL: `${API_URL}/feed-supply`, // URL для отримання всіх поставок корму
        GET_BY_ID: (id: number) => `${API_URL}/feed-supply/${id}`, // URL для отримання поставки корму за ідентифікатором
        CREATE: `${API_URL}/feed-supply/create`, // URL для створення нової поставки корму
        DELETE: (id: number) => `${API_URL}/feed-supply/delete/${id}`, // URL для видалення поставки корму за ідентифікатором
        GET_ALL_VMS: `${API_URL}/feed-supply-vms`, // URL для отримання всіх моделей представлення поставок корму
        GET_ALL_VMS_BY_FEED_ID: (id: number) => `${API_URL}/feed-supply/feed/${id}` // URL для отримання всіх моделей представлення поставок корму за ідентифікатором корму
    }

    // Метод для створення нової поставки корму
    async create(item: FeedSupply): Promise<FeedSupply> {
        const response = await axios.post<FeedSupply>(FeedSupplyService.FEED_SUPPLY_URLS.CREATE, item);
        return response.data;
    }

    // Метод для видалення поставки корму за ідентифікатором
    async delete(itemId: number): Promise<void> {
        await axios.delete(FeedSupplyService.FEED_SUPPLY_URLS.DELETE(itemId));
    }

    // Метод для оновлення існуючої поставки корму (заборонений)
    async update(_item: FeedSupply): Promise<FeedSupply> {
        throw new Error("Method is forbidden.");
    }

    // Метод для отримання всіх поставок корму
    async getAll(): Promise<FeedSupply[]> {
        const response = await axios.get<FeedSupply[]>(FeedSupplyService.FEED_SUPPLY_URLS.GET_ALL);
        return response.data;
    }

    // Метод для отримання поставки корму за ідентифікатором
    async getById(id: number): Promise<FeedSupply | null> {
        const response = await axios.get<FeedSupply | null>(FeedSupplyService.FEED_SUPPLY_URLS.GET_BY_ID(id));
        return response.data;
    }

    // Метод для отримання всіх моделей представлення поставок корму
    async getAllVMs(): Promise<FeedSupplyVM[]> {
        const response = await axios.get<[]>(FeedSupplyService.FEED_SUPPLY_URLS.GET_ALL_VMS);
        const vms: FeedSupplyVM[] = [];
        response.data.map(feedSupply => {
            //@ts-ignore
            vms.push(new FeedSupplyVM(feedSupply.id, feedSupply.amount, feedSupply.timestamp, feedSupply.storekeeper_name, feedSupply.storekeeper_surname));
        });
        return vms;
    }

    // Метод для отримання всіх моделей представлення поставок корму за ідентифікатором корму
    async getAllVMsByFeedId(id: number): Promise<FeedSupplyVM[]> {
        const response = await axios.get<[]>(FeedSupplyService.FEED_SUPPLY_URLS.GET_ALL_VMS_BY_FEED_ID(id));
        const vms: FeedSupplyVM[] = [];
        response.data.map(feedSupply => {
            //@ts-ignore
            vms.push(new FeedSupplyVM(feedSupply.id, feedSupply.amount, feedSupply.timestamp, feedSupply.storekeeper_name, feedSupply.storekeeper_surname));
        });
        return vms;
    }
}

File: apzkr-pzpi-21-11-milennyi-ivan/Task1-Server/src/db/services/feed_supply_service.rs
use std::sync::Arc;
use async_trait::async_trait;
use sqlx::{MySql, Pool, query, query_as};
use crate::db::service_error::ServiceError;
use crate::db::traits::{FeedSupplyManage, Service};
use crate::models::FeedSupply;
use crate::view_models::FeedSupplyVM;

// Сервіс для управління постачанням кормів
pub(crate) struct FeedSupplyService<T> {
    pool: Arc<T>,
}

#[async_trait]
impl Service<Pool<MySql>> for FeedSupplyService<Pool<MySql>> {
    type Model = FeedSupply;
    type Error = ServiceError;
    type ViewModel = FeedSupplyVM;

    // Функція для створення нового сервісу
    fn new(pool: Arc<Pool<MySql>>) -> Self {
        FeedSupplyService { pool }
    }

    // Функція для створення нового постачання корму
    async fn create(&self, mut item: Self::Model) -> Result<Self::Model, Self::Error> {
        let mut transaction = self.pool.begin().await.map_err(|error| ServiceError::DatabaseError(error))?;

        query(
            r#"
            UPDATE Feeds
            SET amount = amount + ?
            WHERE id = ?
            "#
        )
            .bind(item.amount())
            .bind(item.feed_id())
            .execute(&mut *transaction).await
            .map_err(|error| ServiceError::DatabaseError(error))?;

        let result = query(
            r#"
            INSERT INTO FeedSupplies (storekeeper_id, amount, timestamp, feed_id)
            VALUES (?, ?, ?, ?)
            "#
        )
            .bind(item.storekeeper_id())
            .bind(item.amount())
            .bind(item.timestamp())
            .bind(item.feed_id())
            .execute(&mut *transaction).await
            .map_err(|error| ServiceError::DatabaseError(error))
            .map(|result|
                if result.rows_affected() == 1 {
                    item.set_id(result.last_insert_id());
                    Ok(item)
                } else {
                    Err(ServiceError::CustomError("Insertion went wrong. Zero rows affected".to_string()))
                }
            )
            .unwrap_or_else(|error| Err(error));

        transaction.commit().await.map_err(|error| ServiceError::DatabaseError(error))?;

        result
    }

    // Функція для видалення постачання корму за його ідентифікатором
    async fn delete(&self, item_id: u64) -> Result<(), Self::Error> {
        query(
            r#"
            DELETE FROM FeedSupplies
            WHERE id = ?
            "#
        )
            .bind(item_id)
            .execute(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
            .map(|result|
                if result.rows_affected() == 0 {
                    Err(ServiceError::CustomError("Zero rows affected".to_string()))
                } else {
                    Ok(())
                }
            )
            .unwrap_or_else(|error| Err(error))
    }

    // Функція для оновлення інформації про постачання корму (заборонено)
    async fn update(&self, _item: Self::Model) -> Result<Self::Model, Self::Error> {
        Err(ServiceError::ForbiddenError)
    }

    // Функція для отримання всіх постачань кормів
    async fn get_all(&self) -> Result<Vec<Self::Model>, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT id, storekeeper_id, amount, timestamp, feed_id
            FROM FeedSupplies
            "#
        )
            .fetch_all(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
    }

    // Функція для отримання постачання корму за його ідентифікатором
    async fn get_by_id(&self, id: u64) -> Result<Option<Self::Model>, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT id, storekeeper_id, amount, timestamp, feed_id
            FROM FeedSupplies
            WHERE id = ?
            "#
        )
            .bind(id)
            .fetch_optional(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
    }
}

#[async_trait]
impl FeedSupplyManage<Pool<MySql>> for FeedSupplyService<Pool<MySql>> {
    // Функція для отримання всіх ViewModel постачань кормів
    async fn get_all_vms(&self) -> Result<Vec<Self::ViewModel>, Self::Error> {
        query_as::<_, FeedSupplyVM>(
            r#"
            SELECT
            fs.id,
            fs.amount,
            fs.timestamp,
            s.name AS storekeeper_name,
            s.surname AS storekeeper_surname
            FROM FeedSupplies fs
            LEFT JOIN Storekeepers s ON fs.storekeeper_id = s.id
            ORDER BY fs.timestamp DESC
            "#
        )
            .fetch_all(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
    }
    // Функція для отримання всіх ViewModel постачань кормів за ідентифікатором корму
    async fn get_all_vms_by_feed_id(&self, id: u64) -> Result<Vec<Self::ViewModel>, Self::Error> {
        query_as::<_, FeedSupplyVM>(
            r#"
            SELECT
            fs.id,
            fs.amount,
            fs.timestamp,
            s.name AS storekeeper_name,
            s.surname AS storekeeper_surname
            FROM FeedSupplies fs
            LEFT JOIN Storekeepers s ON fs.storekeeper_id = s.id
            WHERE fs.feed_id = ?
            ORDER BY fs.timestamp DESC
            "#
        )
            .bind(id)
            .fetch_all(&*self.pool).await
            .map_err(|error| ServiceError::DatabaseError(error))
    }
}
